### PR TITLE
[client-v2,jdbc-v2] Fix Writing Date/Time values 

### PR DESCRIPTION
**Pull request by chernser**
## Summary

- Fixes writing date/time values 
- Fixes timezone conversion

Closes https://github.com/ClickHouse/clickhouse-java/issues/2701
Closes https://github.com/ClickHouse/clickhouse-java/issues/2065
Closes https://github.com/ClickHouse/clickhouse-java/issues/2496
Closes https://github.com/ClickHouse/clickhouse-java/issues/1220
Closes https://github.com/ClickHouse/clickhouse-java/issues/1117
Closes https://github.com/ClickHouse/clickhouse-java/issues/1048
Closes https://github.com/ClickHouse/clickhouse-java/issues/2381
Closes https://github.com/ClickHouse/clickhouse-java/issues/1735
Closes https://github.com/ClickHouse/clickhouse-java/issues/2542
Closes https://github.com/ClickHouse/clickhouse-java/issues/2557

## Checklist
Delete items not relevant to your PR:
- [x] Closes #<issue ref>
- [x] Unit and integration tests covering the common scenarios were added
- [x] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

---

> [!NOTE]
> **Medium Risk**
> Touches core JDBC parameter encoding and `ResultSet` date/time conversion paths, so regressions could surface as subtle timezone/day-shift or precision issues across environments. Changes are well-covered with expanded unit/integration tests but affect widely used APIs.
> 
> **Overview**
> Fixes JDBC/client-v2 handling of `Date`/`Time`/`Timestamp` values to be *timezone-explicit* and avoid day-shift/precision bugs when using `Calendar` or non-default zones.
> 
> `DataTypeUtils` adds bidirectional SQL↔java.time conversion helpers plus safe nanosecond unix-timestamp string generation (overflow-safe) and improved time/datetime formatters. JDBC `PreparedStatementImpl`, `WriterStatementImpl`, and `ResultSetImpl` are updated to consistently use `defaultCalendar` (instead of `null`) and these new utilities; parameter encoding now supports `Duration` for `Time64`, encodes `LocalDateTime`/`Instant` via `fromUnixTimestamp64Nano(...)`, and validates `setObject(..., targetSqlType)` types instead of emitting implicit `CAST`s.
> 
> Client record readers (`AbstractBinaryFormatReader`, `MapBackedRecord`) now allow `Time/Time64` values read as `LocalDateTime` to be retrieved as `Duration`. Tests are heavily expanded/adjusted to cover timezone round-trips and day-shift edge cases; several implicit type-cast tests are disabled as invalid behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4ae6fdf8c85295bb2268639f3f607877eba66c43. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/PreparedStatementImpl.java [hunk 1]
```diff
@@ -262,27 +255,31 @@ int getParametersCount() {
...
+
+        isValidForTargetType(x, targetSqlType.getVendorTypeNumber());
+        values[parameterIndex-1] = encodeObject(x);
     }
 
     @Override
     public void setObject(int parameterIndex, Object x, SQLType targetSqlType, int scaleOrLength) throws SQLException {
         ensureOpen();
-        values[parameterIndex-1] = encodeObject(x, sqlType2ClickHouseDataType(targetSqlType), scaleOrLength);
+        values[parameterIndex-1] = encodeObject(x, (long) scaleOrLength);
```
[the remark this answers is not in the document]
Comment (chernser) [reply jdbc-v2/src/main/java/com/clickhouse/jdbc/PreparedStatementImpl.java:284]: added check.

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/PreparedStatementImpl.java [hunk 2]
```diff
@@ -1006,45 +981,36 @@ private static String encodeCharacterStream(Reader reader, Long length) throws S
...
+            case Types.TIME_WITH_TIMEZONE:
+                if (vClass == LocalTime.class || vClass == java.sql.Time.class || vClass == LocalDateTime.class || vClass == ZonedDateTime.class) {
+                    return;
+                }
+                break;
+            case Types.TIMESTAMP:
+            case Types.TIMESTAMP_WITH_TIMEZONE:
+                if (vClass == Timestamp.class || vClass == LocalDateTime.class || vClass == ZonedDateTime.class) {
+                    return;
+                }
```
[the remark this answers is not in the document]
Comment (chernser) [reply jdbc-v2/src/main/java/com/clickhouse/jdbc/PreparedStatementImpl.java:1013]: done.

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/ResultSetImpl.java [hunk 3]
```diff
@@ -1015,17 +1021,15 @@ public Date getDate(int columnIndex, Calendar cal) throws SQLException {
...
                 wasNull = true;
                 return null;
             }
             wasNull = false;
 
-            Calendar c = (Calendar) (cal != null ? cal : defaultCalendar).clone();
-            c.clear();
-            c.set(ld.getYear(), ld.getMonthValue() - 1, ld.getDayOfMonth(), 0, 0, 0);
-            return new Date(c.getTimeInMillis());
+            return DataTypeUtils.toSqlDate(ld, cal.getTimeZone());
```
[the remark this answers is not in the document]
Comment (chernser) [reply jdbc-v2/src/main/java/com/clickhouse/jdbc/ResultSetImpl.java:1032]: done. fix all methods.

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/WriterStatementImpl.java [hunk 4]
```diff
@@ -47,12 +48,14 @@ public class WriterStatementImpl extends PreparedStatementImpl implements Prepar
...
     private ClickHouseBinaryFormatWriter writer;
     private final TableSchema tableSchema;
+    private final Calendar defaultCalendar;
 
     public WriterStatementImpl(ConnectionImpl connection, String originalSql, TableSchema tableSchema,
                                ParsedPreparedStatement parsedStatement)
             throws SQLException {
         super(connection, originalSql, parsedStatement);
 
+        this.defaultCalendar = connection.getDefaultCalendar();
```
[the remark this answers is not in the document]
Comment (chernser) [reply jdbc-v2/src/main/java/com/clickhouse/jdbc/WriterStatementImpl.java]: fixed.

File: client-v2/src/test/java/com/clickhouse/client/api/DataTypeUtilsTests.java [hunk 5]
```diff
@@ -130,4 +134,579 @@ void formatInstantForDateTime64Truncated() {
...
+    @Test(groups = {"unit"})
+    void testDifferentDateConversions() throws Exception {
+        Calendar externalSystemTz = Calendar.getInstance(TimeZone.getTimeZone("UTC+12"));
+        Calendar utcTz = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        Calendar applicationLocalTz = Calendar.getInstance(TimeZone.getTimeZone("UTC-8"));
+
+
+        String externalDateStr = externalSystemTz.get(Calendar.YEAR) + "-" + (externalSystemTz.get(Calendar.MONTH) + 1)  + "-" + externalSystemTz.get(Calendar.DAY_OF_MONTH);
+        java.sql.Date externalDate = new java.sql.Date(externalSystemTz.getTimeInMillis());
+        System.out.println(externalDate.toLocalDate());
```
Comment (mzitnik) [inline client-v2/src/test/java/com/clickhouse/client/api/DataTypeUtilsTests.java]: Should we clean `System.out`

File: jdbc-v2/src/main/java/com/clickhouse/jdbc/PreparedStatementImpl.java [hunk 6]
```diff
@@ -469,43 +468,22 @@ public static String replaceQuestionMarks(String sql, final String replacement)
...
-        return c.toInstant();
+        TimeZone tz = (cal == null ? defaultCalendar : cal).getTimeZone();
+        values[parameterIndex - 1] = encodeObject(DataTypeUtils.toLocalDate(x, tz));
     }
 
     @Override
     public void setTime(int parameterIndex, Time x, Calendar cal) throws SQLException {
         ensureOpen();
-        values[parameterIndex - 1] = encodeObject(sqlTimeToInstant(x, cal));
-    }
```
[the remark this answers is not in the document]
Comment (chernser) [reply jdbc-v2/src/main/java/com/clickhouse/jdbc/PreparedStatementImpl.java:487]: the reinterpretation done in `toZonedDateTime`
this is tested. 

Besides your comment is off by some lines.
